### DISCUSSION TITLE
[Feature] YARA scanner module

### DISCUSSION THREAD

**Discussion by dspruell** [rspamd/rspamd#3511] [Ideas]
## Summary

Add YARA scanner module. YARA is a flexible and performant file scanning engine and library that supports rules for detecting content within files. A YARA module for Rspamd would function in many ways similar to the antivirus module.

## Motivation

YARA is an established standard within the antivirus and malware research and response field. Many users have their own rules written for YARA, security vendors supply YARA rulesets, and there are public repositories of YARA rules written for the detection of malware and other use cases. YARA module support would enable Rspamd to natively integrate and be extended to scan received emails and attachments with YARA. The expected outcome is that Rspamd would be instantly useful to a number of users who use YARA and wish to implement it at their email gateway to rapidly respond to malware laden spam deliveries.

## Describe alternatives you've considered

Rspamd has existing customizability, but YARA is wonderful enough to justify dedicated support. YARA handles advanced use cases in a performant manner including byte jumps, available modules and other support that are difficult to express in regular expressions and is tedious to reimplement in Lua. YARA enables rules to contain tags and metadata and optionally includes both of these in match output, providing rich context that would be available for symbols and log data.

A potential alternative to this is to use the External Services module with an ICAP server. Two options for this exist, neither particularly excellent (require maintaining an ICAP server/require maintaining custom modules/unmaintained):

* https://github.com/fygrave/c_icap_yara
* https://github.com/RamadhanAmizudin/python-icap-yara

Rudimentary YARA support _is_ available through [ClamAV](https://www.clamav.net/documents/using-yara-rules-in-clamav) today, however it is laden with caveats. Because of this, use cases are limited and available rulesets are likely to be partially unusable compared to a more direct integration.

**Comment by vstakhov** [comment]
I have checked yara in the past and found nothing useful comparing to general regular expressions with Hyperscan. Furthermore, with the recent Rspamd version, Lua postfilter feature provides a very flexible and fast way to utilise Hyperscan performance with flexibility of Lua logic

**Comment by vstakhov** [comment]
Maybe that's not the best wording: I think yara support is a good thing. But I personally have neither time nor requests to implement Yara in Rspamd. Using of side libraries for that is almost 100% a bad idea when we have Hyperscan here.
